### PR TITLE
Make index_add take a scalar argument alpha

### PR DESCRIPTION
**Pull request by zasdfgbnm**
```
index_add(Tensor self, int dim, Tensor index, Tensor source) -> Tensor
```
now becomes
```
index_add(Tensor self, int dim, Tensor index, Tensor source, Scalar alpha=1) -> Tensor
```
Generally, this sounds useful and harmless, and inside PyTorch, we are already needing this feature in `add_out_dense_sparse_cuda`, see the `SparseCUDATensorMath.cu` change in this PR.

**Test not added yet. Will add if after discussion we believe this is a good idea.**
- [ ] TODO: add test

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
cc: @ptrblck

**Review by albanD — COMMENTED** [review]
I am confused, the code is changing `index_add` and `index_add_` but you mention `index_put` in the description?

Also what is this alpha parameter doing? Is the goal here just to fuse an element-wise multiplication inside the kernel?

**Comment by zasdfgbnm** [comment]
@albanD Yes, sorry about the confusion, I mistakenly typed `index_put` which is wrong and should be `index_add`. And yes, you are right. It is just to fuse an elementwise multiplication inside the kernel

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -558,7 +558,7 @@
 - name: index.Tensor(Tensor self, Tensor?[] indices) -> Tensor
   self: index_backward(grad.new_zeros(self.sizes(), self.options()), indices, grad)
 
-- name: index_add_(Tensor(a!) self, int dim, Tensor index, Tensor source) -> Tensor(a!)
+- name: index_add_(Tensor(a!) self, int dim, Tensor index, Tensor source, Scalar alpha=1) -> Tensor(a!)
   self: grad
   source: grad.index_select(dim, index).expand_as(source)
```
Comment (ngimel) [inline tools/autograd/derivatives.yaml:563]: maybe_multiply(..., alpha)

Comment (albanD) [reply tools/autograd/derivatives.yaml:563]: Yes that's why I was asking about the role.
This formula definitely needs to be updated as natalia suggested.
I think we also want the test and doc to be updated.

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -5005,16 +5005,16 @@
     CPU: legacy::cpu::_th_put_
     CUDA: legacy::cuda::_th_put_
 
-- func: index_add_(Tensor(a!) self, int dim, Tensor index, Tensor source) -> Tensor(a!)
+- func: index_add_(Tensor(a!) self, int dim, Tensor index, Tensor source, Scalar alpha=1) -> Tensor(a!)
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:5008]: Suggest make alpha kwarg-only

**Review by mruberry — CHANGES_REQUESTED** [review]
As @ngimel and @albanD mention this needs to:

- update the docs
- add a test
- fix the autograd formula

It should also:

- create new signatures to avoid a forward-compatibility issue

To validate that autograd with alpha works properly the sample inputs for index_add can be extended here:

https://github.com/pytorch/pytorch/blob/635595f706f83b9f4bf3193afffdbf297e1d6422/torch/testing/_internal/common_methods_invocations.py#L725

File: tools/autograd/derivatives.yaml (hunk 1 above)
Comment (zasdfgbnm) [reply tools/autograd/derivatives.yaml]: fixed

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/native_functions.yaml]: fixed

**Comment by zasdfgbnm** [comment]
@albanD @ngimel @mruberry I have resolved all review comments.

**Review by mruberry — APPROVED** [review]
LGTM. @albanD, @ngimel, would you like to take another look?

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**